File: lib/valuation.ts
/**
 * The valuation layer the screener is built on.
 *
 * Every method is wired to real math in `api/valuation.py` and scored across
 * the index by `POST /backfill/valuations`. All report the same contract — a
 * fair value and a confidence — so the UI never needs to know which model
 * produced a number.
 *
 * A model that cannot speak to a company emits NO verdict rather than a zero,
 * which is why `verdicts` is ragged and coverage is worth filtering on.
 *
 * CAPM and Fama-French are deliberately absent: they produce an expected
 * return, not a value per share. They appear in the discount-rate panel.
 */

export type MethodId = "fcfe" | "fcff" | "comps" | "ddm" | "rim"

export type MethodStatus = "live" | "planned"

export type ValuationMethod = {
  id: MethodId
  label: string
  /** Expanded name, shown once in the methodology panel. */
  full: string
  status: MethodStatus
  /** One line the user can act on — what the method rewards. */
  blurb: string
  /** The core expression, in the notation used in the Python implementation. */
  formula: string
  /**
   * The conditions under which this model declines to produce a verdict.
   * Mirrors the guards in `api/valuation.py`; kept here so the UI can explain
   * an absent row rather than leaving it looking like missing data.
   */
  refusesWhen: string[]
}

export const VALUATION_METHODS: ValuationMethod[] = [
  {
    id: "fcfe",
    label: "FCFE",
    full: "Free Cash Flow to Equity",
    status: "live",
    blurb:
      "Cash left for shareholders after capex and net borrowing, discounted at the cost of equity. Skipped for banks and REITs, where capex is not a meaningful concept.",
    formula: "FCFE = CFO + CapEx + net borrowing, discounted at ke",
    refusesWhen: [
      "Trailing free cash flow to equity is zero or negative",
      "Banks and REITs, where capital expenditure is not a meaningful concept",
      "Fewer than 4 quarters of statements on file",
      "Cost of equity within 2 points of terminal growth",
    ],
  },
  {
    id: "fcff",
    label: "FCFF",
    full: "Free Cash Flow to Firm",
    status: "live",
    blurb:
      "Cash available to all capital providers, discounted at WACC and bridged to equity by netting out debt. Less sensitive to leverage than FCFE.",
    formula: "FCFF = EBIT(1−t) + D&A + CapEx − ΔWC, discounted at WACC",
    refusesWhen: [
      "Trailing free cash flow to the firm is zero or negative",
      "Banks and REITs",
      "Equity value after netting debt is negative",
      "WACC within 2 points of terminal growth",
    ],
  },
  {
    id: "comps",
    label: "Comps",
    full: "Comparable Company Analysis",
    status: "live",
    blurb:
      "What the company would be worth priced like the median of its sector, on EV/EBITDA and P/E. The one model here that looks at other companies' prices — so it inherits whatever the sector is doing, and will call an expensive company fair if every peer is expensive too.",
    formula: "V₀ = ½·[(median EV/EBITDA × EBITDA − net debt) + median P/E × EPS] ÷ shares",
    refusesWhen: [
      "Negative or zero EBITDA and negative or zero earnings per share",
      "Fewer than 5 usable peers in the sector — a median of a handful is an anecdote",
      "Fewer than 4 quarters of statements on file",
      "Sector unknown, so there is no peer set to compare against",
    ],
  },
  {
    id: "ddm",
    label: "DDM",
    full: "Dividend Discount Model",
    status: "live",
    blurb:
      "Values the dividend stream directly: ten years of explicit dividend growth, then a perpetuity. Skipped below a 1.5% yield, where the dividend explains too little of the price to say anything about value.",
    formula:
      "V₀ = Σ D₀(1 + g)ᵗ / (1 + ke)ᵗ + D₁₀(1 + gₙ) / (ke − gₙ)(1 + ke)¹⁰",
    refusesWhen: [
      "The company pays no dividend",
      "Fewer than 8 dividend payments on record",
      "Yield below 1.5% — a token dividend does not explain the price",
      "Cost of equity within 2 points of the 3% terminal growth rate",
    ],
  },
  {
    id: "rim",
    label: "RIM",
    full: "Residual Income Model",
    status: "live",
    blurb:
      "Book value plus earnings above the cost of equity. Robust where cash flows are lumpy, and the right model for banks — so it carries no sector exclusion.",
    formula: "V₀ = B₀ + Σ (ROE − ke)·B₍ₜ₋₁₎ / (1 + ke)ᵗ",
    refusesWhen: [
      "Negative or zero book value, which makes residual income meaningless",
      "Return on equity unavailable",
    ],
  },
]

/** A single method's verdict on one stock. */
export type MethodVerdict = {
  method: MethodId
  /** Per-share fair value in USD. */
  fairValue: number
  /** Signed fraction: (fairValue - price) / price. Positive = undervalued. */
  marginOfSafety: number
  /** 0–1. How much the inputs support the number. */
  confidence: number
}

/**
 * The rates a company's valuations were discounted at.
 *
 * Not verdicts: a cost of equity is not a value per share, and listing CAPM
 * as a fair-value model left a permanently empty row on every stock page.
 *
 * Every field is nullable — a missing WACC is not a WACC of zero.
 */
export type DiscountRates = {
  /** 13-week T-bill (^IRX), averaged over the same 252-day window. */
  riskFree: number | null
  /** The rate FCFE, DDM and RIM discounted at. */
  costOfEquity: number | null
  /** Which model produced it — the three use incomparable risk premia. */
  costOfEquitySource: "ff5" | "ff3" | "capm" | null
  /** CAPM's own figure, shown beside the one that was used. */
  capmCostOfEquity: number | null
  /** The blended rate FCFF discounted at. */
  wacc: number | null
  costOfDebt: number | null
  /** E / (E + max(net debt, 0)), after the debt-weight cap. */
  equityWeight: number | null
  taxRate: number | null
}

export type Stock = {
  ticker: string
  name: string
  sector: string
  price: number
  marketCap: number
  beta: number
  peRatio: number
  dividendYield: number
  /** Annualised expected return actually realised over the window. */
  realisedReturn: number
  /** Annualised volatility over the same window. */
  volatility: number
  verdicts: MethodVerdict[]
  /** Absent on the offline sample, which has no rates to report. */
  discountRates?: DiscountRates
}

/**
 * The verdicts that count under the current model selection. An empty
 * selection means every model counts, which is the default reading.
 */
export function activeVerdicts(
  stock: Stock,
  methods: MethodId[] = []
): MethodVerdict[] {
  if (methods.length === 0) return stock.verdicts
  return stock.verdicts.filter((v) => methods.includes(v.method))
}

/**
 * Consensus margin of safety, weighted by each method's declared weight (see
 * WEIGHT_* in `api/valuation.py`). The number the screener sorts on.
 *
 * `methods` narrows the consensus to those models only, which is the point of
 * the toggles: picking DDM and FCFE re-answers the whole screen — margins,
 * bands, distribution, ranking — from a cash-returns view alone.
 */
export function consensusMarginOfSafety(
  stock: Stock,
  methods: MethodId[] = []
): number {
  const verdicts = activeVerdicts(stock, methods)
  if (verdicts.length === 0) return 0

  const weight = verdicts.reduce((sum, v) => sum + v.confidence, 0)
  if (weight === 0) return 0

  return (
    verdicts.reduce((sum, v) => sum + v.marginOfSafety * v.confidence, 0) /
    weight
  )
}

export type ValuationBand =
  | "deep-value"
  | "undervalued"
  | "fair"
  | "overvalued"
  | "expensive"
  // No model could value this company. Distinct from "fair" on purpose: a
  // consensus of 0 because nothing applies is not the same claim as a
  // consensus of 0 because the models agree the price is right.
  | "unrated"

/**
 * Whether any of the selected models valued this stock. Under a narrowed
 * selection a stock can be unrated here while still carrying verdicts from
 * models the user switched off — the honest reading: those were asked not to
 * speak.
 *
 * There was a separate agreement floor here, a "valued by at least N models"
 * control. It made the same claim twice: which models are allowed to speak is
 * already the model selection, and a second number modifying it meant two
 * controls had to be reasoned about together to know what a row's band meant.
 * The rating is now exactly the models on screen. Wanting more agreement is
 * expressed by reading the n/m count in the margin column, which was always
 * the more honest place for it — it is per-row, and the floor was not.
 */
export function isRated(stock: Stock, methods: MethodId[] = []): boolean {
  return activeVerdicts(stock, methods).length > 0
}

/** How many models the current selection asks for. Empty means all of them. */
export function selectedModelCount(methods: MethodId[] = []): number {
  return methods.length === 0 ? VALUATION_METHODS.length : methods.length
}

/**
 * Buckets the consensus into the five bands the screener's diverging scale
 * paints.
 *
 * THESE BANDS ARE RELATIVE, NOT ABSOLUTE — quintiles of the actual S&P 500
 * distribution, so "deep value" means "cheapest fifth of the index", not
 * "below intrinsic worth". The UI says so wherever a band is named.
 *
 * That is forced by the models, not chosen: the consensus median sits near
 * -34%, because the cash-flow models discount at ~12% while halving observed
 * growth against a 3% perpetuity. Centring on zero would file three fifths of
 * the index under "expensive" and fail to discriminate within it.
 *
 * If the discount-rate work moves that median, recompute these cut points from
 * the new distribution rather than leaving them frozen here.
 */
export function valuationBand(marginOfSafety: number): ValuationBand {
  if (marginOfSafety >= 0.03) return "deep-value"
  if (marginOfSafety >= -0.26) return "undervalued"
  if (marginOfSafety > -0.41) return "fair"
  if (marginOfSafety > -0.57) return "overvalued"
  return "expensive"
}

/**
 * Shown wherever bands are explained. Kept here beside the cut points so the
 * caveat cannot drift away from the thresholds it describes.
 */
export const BAND_BASIS =
  "Bands are quintiles of the S&P 500 distribution, so they rank a company against the index rather than against its own intrinsic value. The models read the market as expensive almost everywhere — the median consensus is about −34% — so an absolute scale would file three fifths of the index under “expensive”."

/** Shown behind the info trigger on the margin-of-safety filter. */
export const MARGIN_BASIS =
  "Graham’s term for the discount between what a company is worth and what it costs: (fair value − price) ÷ price, averaged over the selected models and weighted by each one’s confidence. +20% means paying 80 cents for a dollar of value. The cushion is the point — buy far enough below fair value and the estimate can be wrong without losing money."

export const BAND_LABELS: Record<ValuationBand, string> = {
  "deep-value": "Deep value",
  undervalued: "Undervalued",
  fair: "Fair value",
  overvalued: "Overvalued",
  expensive: "Expensive",
  unrated: "Unrated",
}

/**
 * Fill colour for a band. Both poles come from the diverging pair in
 * `globals.css`; the midpoint is the neutral gray, never a third hue.
 */
export const BAND_FILL: Record<ValuationBand, string> = {
  "deep-value": "var(--color-undervalued)",
  undervalued: "color-mix(in oklch, var(--color-undervalued), transparent 45%)",
  fair: "var(--color-valuation-neutral)",
  overvalued: "color-mix(in oklch, var(--color-overvalued), transparent 45%)",
  expensive: "var(--color-overvalued)",
  // Deliberately off the diverging scale entirely — "no reading" is not a
  // point on an axis running from cheap to expensive.
  unrated: "var(--color-muted)",
}

export type ScreenerFilters = {
  search: string
  sectors: string[]
  bands: ValuationBand[]
  /**
   * Restrict the consensus to these models. Empty means every model that
   * produced a verdict counts — the default reading.
   */
  methods: MethodId[]
  /** Inclusive [min, max] on the consensus margin of safety. */
  marginRange: [number, number]
  maxBeta: number
}

/**
 * The screen the app opens on.
 *
 * It used to open on [-1, 1] and beta <= 3, which is not a default so much as
 * the absence of one: every rated stock passes, and the first thing on screen
 * is the whole index in an app whose argument is that you should not buy the
 * whole index. A screener's opening state is a worked example of the screen it
 * is for, so this one is a value screen.
 *
 * Both numbers are read off the actual distribution rather than picked for
 * roundness. -25% is just inside the `undervalued` band's -26% cut point, so
 * the default keeps the cheaper two quintiles of the index and drops the three
 * the models price as fair or worse. 1.50 beta trims the high-volatility tail
 * without touching the ordinary market-tracking middle — the S&P's own beta is
 * 1 by construction.
 *
 * Everything downstream reads the filtered set, so this is also what the
 * optimiser is handed on the screener's primary call to action.
 */
export const DEFAULT_FILTERS: ScreenerFilters = {
  search: "",
  sectors: [],
  bands: [],
  methods: [],
  marginRange: [-0.25, 1],
  maxBeta: 1.5,
}

export function applyFilters(
  stocks: Stock[],
  filters: ScreenerFilters
): Stock[] {
  const query = filters.search.trim().toUpperCase()

  return stocks.filter((stock) => {
    if (
      query &&
      !stock.ticker.includes(query) &&
      !stock.name.toUpperCase().includes(query)
    ) {
      return false
    }

    if (filters.sectors.length > 0 && !filters.sectors.includes(stock.sector)) {
      return false
    }

    if (stock.beta > filters.maxBeta) return false

    // An unrated stock has no margin to compare, so the numeric filters below
    // would read its 0 as "fair" and let it through every range. It only
    // qualifies when the band filter asks for it explicitly.
    if (!isRated(stock, filters.methods)) {
      return filters.bands.includes("unrated")
    }

    const margin = consensusMarginOfSafety(stock, filters.methods)
    if (margin < filters.marginRange[0] || margin > filters.marginRange[1]) {
      return false
    }

    if (
      filters.bands.length > 0 &&
      !filters.bands.includes(valuationBand(margin))
    ) {
      return false
    }

    return true
  })
}

File: python-api/valuation.py
"""The valuation models.

Each model is a pure function over plain inputs, returning a verdict dict or
None. Returning None is a first-class outcome and the reason the models are
trustworthy: a bank has no meaningful free cash flow, a non-payer has no
dividend stream, a loss-maker has no Graham number. Rows are only written
for models that produced something, so absence in the UI means "does not
apply here" rather than "fair value is zero".

Factor models are not verdicts. FF3/FF5 produce an expected return, not an
intrinsic value per share. What survives is their correct use: supplying the
cost of equity that the cash-flow models discount at.
"""

from __future__ import annotations

import math

import numpy as np
import pandas as pd

TRADING_DAYS = 252

# Cost of equity is clamped before it is ever used as a discount rate. An
# unclamped ke that drifts near the terminal growth rate makes the Gordon
# denominator collapse and fair values explode into the millions.
MIN_COST_OF_EQUITY = 0.06
MAX_COST_OF_EQUITY = 0.20
# Minimum spread between discount rate and terminal growth. Below this the
# perpetuity is numerically meaningless and the model must refuse.
MIN_DISCOUNT_SPREAD = 0.02
# Long-run nominal growth: roughly real GDP plus inflation. Held below every
# discount rate by MIN_DISCOUNT_SPREAD.
TERMINAL_GROWTH = 0.03
# A 5-year explicit window suits mature businesses but understates companies
# still compounding fast; 10 years lets more of the growth land inside the
# projection rather than being forced into a GDP-rate perpetuity.
PROJECTION_YEARS = 10
# No company compounds faster than nominal GDP forever.
MAX_GROWTH = 0.06

# Factor regressions need enough overlapping observations to mean anything.
MIN_REGRESSION_OBS = 120

MAX_COST_OF_DEBT = 0.15
MIN_COST_OF_DEBT = 0.02

# WACC is a blend, not a cost of equity. A levered firm's WACC belongs BELOW
# its ke, so clamping it to the ke bounds (as this once did) put a 6% floor
# under every levered discount rate -- CHTR sat four basis points above it,
# meaning its discount rate was very nearly a constant.
MIN_WACC = 0.03
MAX_WACC = 0.20
# Market-value weights make the discount rate a function of the price being
# valued. Capping the debt weight bounds that feedback; see `fcff_verdict`.
# This is a modelling judgement, not a fact about the company.
MAX_DEBT_WEIGHT = 0.60

# DDM values the dividend stream and nothing else. Below ~1.5% yield the
# dividend explains too little of the price for Gordon growth to be a statement
# about value — Apple at 0.35% cannot be justified at any sane discount rate —
# so those names get no DDM verdict rather than a uniformly bearish one.
MIN_DDM_YIELD = 0.015

# How much say each methodology gets in the consensus. These are JUDGEMENTS
# about the models, not measurements of any company: FCFF and FCFE model cash
# generation directly and are trusted most; RIM anchors on book value and reads
# low for asset-light compounders, so it is heard but not loudly. Only the
# per-company adjustments below genuinely vary.
WEIGHT_FCFF = 0.80
WEIGHT_FCFE = 0.75
# Comps answer a different question from the four models above -- "what would
# this be worth priced like its peers" rather than "what is this worth" -- so
# it is heard clearly but never louder than a discounted cash flow. If the
# whole sector is mispriced, comps inherits that mispricing wholesale.
WEIGHT_COMPS = 0.55
WEIGHT_DDM = 0.60
WEIGHT_RIM = 0.35

# A sector median drawn from a handful of names is not a median, it is an
# anecdote. Below this many usable peers the multiple is withheld entirely.
MIN_COMPS_PEERS = 5

# Sectors where capex-based cash-flow models do not apply. Banks fund
# themselves with deposits and debt, so "free cash flow" is not owner earnings;
# REITs are depreciation-heavy and are valued on FFO rather than earnings.
NO_CASHFLOW_MODEL_SECTORS = {"Financial Services", "Real Estate"}


def _finite(value) -> float | None:
    """A usable float, or None. Guards every arithmetic entry point."""
    if value is None:
        return None
    try:
        result = float(value)
    except (TypeError, ValueError):
        return None
    return None if math.isnan(result) or math.isinf(result) else result


def _verdict(
    method: str, fair_value: float, price: float, confidence: float
) -> dict | None:
    """Assemble a verdict, rejecting nonsense fair values.

    A fair value at or below zero, or an order of magnitude from the market
    price in EITHER direction, means the model's assumptions broke. The
    symmetry matters: rejecting only the high side censors the undervalued
    readings and tilts every consensus downward.
    """
    fair = _finite(fair_value)
    if fair is None or fair <= 0 or price <= 0:
        return None
    if fair > price * 10 or fair < price / 10:
        return None

    return {
        "method": method,
        "fair_value": round(fair, 4),
        "margin_of_safety": round((fair - price) / price, 6),
        "confidence": round(max(0.0, min(1.0, confidence)), 4),
        "price_at_calc": round(price, 4),
    }


def _stability(values: list[float | None]) -> float:
    """1 − coefficient of variation, clamped to [0, 1].

    A series that swings wildly quarter to quarter is a weaker basis for
    projection than a steady one, and confidence should say so.
    """
    clean = [v for v in (_finite(v) for v in values) if v is not None]
    if len(clean) < 2:
        return 0.5

    mean = sum(clean) / len(clean)
    if mean == 0:
        return 0.3

    variance = sum((v - mean) ** 2 for v in clean) / len(clean)
    cov = math.sqrt(variance) / abs(mean)
    return max(0.0, min(1.0, 1.0 - cov))


def _two_stage_pv(
    base: float, growth: float, discount: float, terminal: float
) -> float | None:
    """Present value of a cash flow growing at `growth` for PROJECTION_YEARS,
    then at `terminal` in perpetuity."""
    if discount - terminal < MIN_DISCOUNT_SPREAD:
        return None

    pv = 0.0
    flow = base
    for year in range(1, PROJECTION_YEARS + 1):
        flow = flow * (1 + growth)
        pv += flow / ((1 + discount) ** year)

    terminal_value = flow * (1 + terminal) / (discount - terminal)
    pv += terminal_value / ((1 + discount) ** PROJECTION_YEARS)
    return pv


# --------------------------------------------------------------------------
# Factor models
# --------------------------------------------------------------------------


def factor_regression(
    excess_returns: pd.Series, factor_df: pd.DataFrame, columns: list[str]
) -> dict | None:
    """OLS of excess stock returns on factor returns.

    Runs on the date INTERSECTION of the two series — never on the raw last-252
    price dates, because the Fama-French library publishes weeks in arrears and
    assuming coverage would silently regress against misaligned dates.
    """
    joined = pd.concat([excess_returns.rename("y"), factor_df[columns]], axis=1)
    joined = joined.dropna()

    if len(joined) < MIN_REGRESSION_OBS:
        return None

    joined = joined.tail(TRADING_DAYS)

    y = joined["y"].to_numpy(dtype=float)
    x = joined[columns].to_numpy(dtype=float)
    design = np.column_stack([np.ones(len(x)), x])

    try:
        coefficients, *_ = np.linalg.lstsq(design, y, rcond=None)
    except np.linalg.LinAlgError:
        return None

    fitted = design @ coefficients
    residuals = y - fitted
    total_variance = float(np.var(y))
    if total_variance == 0:
        return None

    r_squared = 1.0 - float(np.var(residuals)) / total_variance

    return {
        "alpha_daily": float(coefficients[0]),
        "betas": {name: float(b) for name, b in zip(columns, coefficients[1:])},
        "r_squared": r_squared,
        "n_obs": len(joined),
    }


def factor_cost_of_equity(
    betas: dict[str, float], premia: dict[str, float], risk_free: float
) -> float:
    """Cost of equity implied by factor loadings: rf + sum(beta_i * lambda_i)."""
    total = risk_free + sum(
        beta * premia.get(name, 0.0) for name, beta in betas.items()
    )
    return max(MIN_COST_OF_EQUITY, min(MAX_COST_OF_EQUITY, total))


# --------------------------------------------------------------------------
# Cash-flow and accounting models
# --------------------------------------------------------------------------


def ddm_verdict(
    price: float,
    trailing_dividend: float | None,
    dividend_growth: float | None,
    cost_of_equity: float,
    payment_count: int,
    payout_ratio: float | None,
    payment_stability: float,
) -> dict | None:
    """Two-stage dividend discount: the trailing dividend grown explicitly for
    PROJECTION_YEARS, then at TERMINAL_GROWTH in perpetuity.

    Refuses for non-payers. No dividend is not evidence of zero value; it means
    this model has nothing to say about the company.

    Two-stage rather than single-stage Gordon because the closed form cannot
    represent a dividend growing faster than its own discount rate: its ke > g
    guard refused 109 of the index's 248 qualifying payers.
    """
    dividend = _finite(trailing_dividend)
    if dividend is None or dividend <= 0 or payment_count < 8:
        return None

    # A token dividend does not make a company valuable as a dividend stream.
    if dividend / price < MIN_DDM_YIELD:
        return None

    # Halved on the way in for the reason the cash-flow models halve revenue
    # growth: one lookback window's rate is not a decade's rate. It matters more
    # here, because a dividend reinstatement reads as a triple-digit CAGR that
    # describes an event rather than a trend.
    growth = min(MAX_GROWTH, max(0.0, _finite(dividend_growth) or 0.0) * 0.5)

    fair_value = _two_stage_pv(dividend, growth, cost_of_equity, TERMINAL_GROWTH)
    if fair_value is None:
        return None

    # DDM only speaks for companies that genuinely return value as dividends,
    # and MIN_DDM_YIELD has already excluded the ones it cannot describe.
    confidence = WEIGHT_DDM * payment_stability
    payout = _finite(payout_ratio)
    # A payout above earnings is being funded from somewhere other than
    # profit, so the stream is less durable than its history suggests.
    if payout is not None and payout > 0.9:
        confidence *= 0.6

    return _verdict("ddm", fair_value, price, confidence)


def fcfe_verdict(
    price: float,
    ttm_operating_cf: float | None,
    ttm_capex: float | None,
    ttm_net_borrowing: float | None,
    shares: float | None,
    cost_of_equity: float,
    growth: float,
    quarters_used: int,
    sector: str | None,
) -> dict | None:
    """Free cash flow to equity, discounted at the cost of equity.

    FCFE = operating cash flow + capex + net borrowing. Capex is stored
    negative, so this is an addition.
    """
    if sector in NO_CASHFLOW_MODEL_SECTORS:
        return None
    if quarters_used < 4:
        return None

    ocf = _finite(ttm_operating_cf)
    capex = _finite(ttm_capex)
    share_count = _finite(shares)
    if ocf is None or capex is None or share_count is None or share_count <= 0:
        return None

    fcfe = ocf + capex + (_finite(ttm_net_borrowing) or 0.0)
    if fcfe <= 0:
        return None

    # Spread is checked against the requested growth before clamping, so an
    # unreasonable growth estimate refuses rather than being capped into
    # looking reasonable.
    if cost_of_equity - max(0.0, growth) < MIN_DISCOUNT_SPREAD:
        return None

    equity_value = _two_stage_pv(
        fcfe, min(growth, MAX_GROWTH), cost_of_equity, TERMINAL_GROWTH
    )
    if equity_value is None:
        return None

    confidence = WEIGHT_FCFE
    # Cash flow propped up by borrowing is not owner earnings.
    borrowing = _finite(ttm_net_borrowing) or 0.0
    if borrowing > 0.5 * fcfe:
        confidence *= 0.6

    return _verdict("fcfe", equity_value / share_count, price, confidence)


def wacc_components(
    equity_value: float,
    tax_rate: float,
    total_debt: float | None,
    net_debt: float | None,
    ttm_interest: float | None,
    risk_free: float,
    cost_of_equity: float,
) -> dict:
    """Every input to WACC, and the resulting rate, as one dict.

    Extracted so the discount-rate panel shows exactly the numbers the DCF
    discounted at. A second implementation written for display is how a UI ends
    up quoting a WACC the model never used.

    `equity_value` is price x shares, measured by the caller so it agrees with
    the price the verdict is scored against.
    """
    # Cost of debt is what the company actually pays on money it has borrowed,
    # so it is a yield on GROSS debt. Netting cash out here would invent a
    # borrowing rate no lender offers.
    gross_debt = _finite(total_debt) or 0.0
    interest = _finite(ttm_interest)
    if gross_debt > 0 and interest is not None and interest > 0:
        cost_of_debt = interest / gross_debt
        imputed_debt_cost = False
    else:
        cost_of_debt = risk_free + 0.02
        imputed_debt_cost = True
    cost_of_debt = max(MIN_COST_OF_DEBT, min(MAX_COST_OF_DEBT, cost_of_debt))

    # The weights use NET debt, because the equity bridge subtracts net debt.
    # Gross weights against a net bridge counts a cash pile twice. Net cash
    # floors at zero: more cash than debt is a 100% equity weight, not a
    # subsidy.
    bridge_debt = _finite(net_debt) or 0.0
    weight_debt = max(0.0, bridge_debt)

    total_capital = equity_value + weight_debt
    raw_debt_weight = weight_debt / total_capital if total_capital > 0 else 0.0
    # Market-value weights make the discount rate a function of the very price
    # being valued: as equity falls its weight falls, the cheap after-tax debt
    # weight rises, WACC drops, and the model calls the falling equity MORE
    # valuable. Capping the debt weight bounds that loop.
    debt_weight = min(raw_debt_weight, MAX_DEBT_WEIGHT)

    wacc = (1 - debt_weight) * cost_of_equity + debt_weight * cost_of_debt * (
        1 - tax_rate
    )
    wacc = max(MIN_WACC, min(MAX_WACC, wacc))

    return {
        "wacc": wacc,
        "cost_of_debt": cost_of_debt,
        "imputed_debt_cost": imputed_debt_cost,
        "tax_rate": tax_rate,
        "equity_value": equity_value,
        "net_debt": bridge_debt,
        "debt_weight": debt_weight,
        "equity_weight": 1 - debt_weight,
        "debt_weight_capped": raw_debt_weight > MAX_DEBT_WEIGHT,
    }


def fcff_verdict(
    price: float,
    ttm_ebit: float | None,
    ttm_depreciation: float | None,
    ttm_capex: float | None,
    ttm_working_capital: float | None,
    tax_rate: float | None,
    total_debt: float | None,
    net_debt: float | None,
    ttm_interest: float | None,
    shares: float | None,
    risk_free: float,
    cost_of_equity: float,
    growth: float,
    quarters_used: int,
    sector: str | None,
) -> dict | None:
    """Free cash flow to the firm, discounted at WACC, bridged to equity."""
    if sector in NO_CASHFLOW_MODEL_SECTORS:
        return None
    if quarters_used < 4:
        return None

    ebit = _finite(ttm_ebit)
    capex = _finite(ttm_capex)
    share_count = _finite(shares)
    if ebit is None or capex is None or share_count is None or share_count <= 0:
        return None

    # Measured from the same price the margin of safety uses, times the same
    # share count the fair value is divided by. `company_profile.market_cap` is
    # a yfinance snapshot from a different backfill and disagreed by 13% on
    # CHTR; one number for the weights and another for the division cannot both
    # be right.
    equity_mv = price * share_count
    if equity_mv <= 0:
        return None

    rate = _finite(tax_rate)
    imputed_tax = rate is None
    if imputed_tax:
        rate = 0.21
    rate = max(0.0, min(0.5, rate))

    depreciation = _finite(ttm_depreciation) or 0.0
    working_capital = _finite(ttm_working_capital) or 0.0
    fcff = ebit * (1 - rate) + depreciation + capex - working_capital
    if fcff <= 0:
        return None

    components = wacc_components(
        equity_mv, rate, total_debt, net_debt, ttm_interest, risk_free,
        cost_of_equity,
    )
    wacc = components["wacc"]
    bridge_debt = components["net_debt"]

    # As in FCFE: validate the spread against the requested growth first.
    if wacc - max(0.0, growth) < MIN_DISCOUNT_SPREAD:
        return None

    enterprise_value = _two_stage_pv(
        fcff, min(growth, MAX_GROWTH), wacc, TERMINAL_GROWTH
    )
    if enterprise_value is None:
        return None

    equity_value = enterprise_value - bridge_debt
    if equity_value <= 0:
        return None

    confidence = WEIGHT_FCFF
    if imputed_tax:
        confidence *= 0.8

    return _verdict("fcff", equity_value / share_count, price, confidence)


def comps_verdict(
    price: float,
    ttm_ebitda: float | None,
    ttm_eps: float | None,
    net_debt: float | None,
    shares: float | None,
    sector_multiples: dict[str, float] | None,
    quarters_used: int,
) -> dict | None:
    """Relative valuation against the median multiples of the company's sector.

    Answers "what would this be worth priced like its peers", a different
    question from the cash-flow models and the reason it sits beside them. It
    inherits whatever the sector is doing: if every peer is expensive, comps
    calls an expensive company fair.

    Averages the values implied by EV/EBITDA and P/E rather than preferring
    one. They fail in different places, so the average is steadier than either.
    """
    if quarters_used < 4 or not sector_multiples:
        return None

    share_count = _finite(shares)
    if share_count is None or share_count <= 0:
        return None

    implied: list[float] = []

    # EV/EBITDA -> enterprise value -> equity, by netting debt back out.
    ev_multiple = sector_multiples.get("ev_ebitda")
    ebitda = _finite(ttm_ebitda)
    if ev_multiple is not None and ebitda is not None and ebitda > 0:
        equity_value = ev_multiple * ebitda - (_finite(net_debt) or 0.0)
        if equity_value > 0:
            implied.append(equity_value / share_count)

    # P/E lands directly on a per-share value.
    pe_multiple = sector_multiples.get("pe")
    eps = _finite(ttm_eps)
    if pe_multiple is not None and eps is not None and eps > 0:
        implied.append(pe_multiple * eps)

    if not implied:
        return None

    return _verdict("comps", sum(implied) / len(implied), price, WEIGHT_COMPS)


def rim_verdict(
    price: float,
    book_value_ps: float | None,
    eps: float | None,
    cost_of_equity: float,
    payout_ratio: float | None,
    quarters_used: int,
) -> dict | None:
    """Residual income: book value plus earnings above the cost of equity.

    Robust where cash flows are lumpy, and the right model for banks — which is
    why it carries no sector exclusion. Refuses on negative book value, which
    makes the model meaningless and does occur among buyback-heavy large caps.
    """
    book = _finite(book_value_ps)
    earnings = _finite(eps)
    if book is None or earnings is None or book <= 0 or quarters_used < 4:
        return None

    payout = _finite(payout_ratio)
    retention = 1.0 - (payout if payout is not None and 0 <= payout <= 1 else 0.4)

    value = book
    carried_book = book
    for year in range(1, PROJECTION_YEARS + 1):
        residual = earnings - cost_of_equity * carried_book
        value += residual / ((1 + cost_of_equity) ** year)
        carried_book += earnings * retention

    # Like EPV, RIM anchors on book value and credits no growth beyond
    # retained earnings, so it reads low for asset-light compounders whose
    # value is not on the balance sheet. Weighted accordingly.
    return _verdict(
        "rim", value, price, WEIGHT_RIM * min(1.0, quarters_used / 4)
    )


def damp_cashflow_confidence(verdicts: list[dict]) -> list[dict]:
    """Scale down confidence when both cash-flow models agree.

    FCFE and FCFF discount the same underlying cash generation by two different
    routes. Left alone they would give one methodology two votes in a
    weighted consensus, so each is scaled by 1/sqrt(n) when both fire.
    """
    cashflow_methods = {"fcfe", "fcff"}
    firing = [v for v in verdicts if v["method"] in cashflow_methods]
    if len(firing) <= 1:
        return verdicts

    scale = 1.0 / math.sqrt(len(firing))
    for verdict in firing:
        verdict["confidence"] = round(verdict["confidence"] * scale, 4)
    return verdicts
